iotconnect-sdk/nrf-layer-lib/src/iotconnect_mqtt.c

```diff
@@ -50,10 +69,141 @@ static int fds_init(struct mqtt_client *c);
+
+    msg_wait_puback_list_t *p = msg_wait_puback_list_head.p_next;
+
+    while (p) {
+        if (p->msg_wait_puback.msg_id && (p->msg_wait_puback.msg_id == msg_id)) {
+            p->p_prev->p_next = p->p_next;
+            if (p->p_next) {
+                p->p_next->p_prev = p->p_prev;
+            }
+            free(p);
+            //printk("Message Id \"%u\" deleted from list\n", msg_id);
+            return true;
+        } else {
+            p = p->p_next;
+        }
+    }
+    return false;
+}
+
+/********************************************************************/
+/* Function to clear the list.                                      */
+/********************************************************************/
+static void msg_wait_puback_list_clear(void) {
+
+    msg_wait_puback_list_t *p = msg_wait_puback_list_head.p_next;
+
+    while (msg_wait_puback_list_head.p_next) {
+        p = msg_wait_puback_list_head.p_next;
+        p->p_prev->p_next = p->p_next;
+        if (p->p_next) {
+            p->p_next->p_prev = p->p_prev;
+        }
+        free(p);
+    }
+
+    //printk("list cleared.\n");
+
+}
+
+/********************************************************************/
+/* Function to find msg ID from the list.                           */
+/********************************************************************/
+static bool msg_wait_puback_list_find(uint32_t msg_id) {
+
+    msg_wait_puback_list_t *p = msg_wait_puback_list_head.p_next;
+    bool found = false;
+
+    while (p) {
+        if (p->msg_wait_puback.msg_id && (p->msg_wait_puback.msg_id == msg_id)) {
+            found = true;
+            break;
+        }
+        p = p->p_next;
+    }
+
+    if (!found) {
+        printk("Message Id \"%u\" NOT found in list\n", msg_id);
+    }
+
+    return found;
+
+}
+
+/********************************************************************/
+/* Function to handle timeout of all msg ID in the list.            */
+/********************************************************************/
+static void msg_wait_puback_list_timeout_handling_proc(void) {
+
+    msg_wait_puback_list_t *p = msg_wait_puback_list_head.p_next;
+    uint32_t msg_id;
+
+    while(p){
+        if (p->msg_wait_puback.msg_id &&
+            ((k_uptime_get_32() - p->msg_wait_puback.start_ticks) >= (config->puback_timeout_cfg.timeout_s * 1000))) {
+
+            printk("Message Id \"%d\" timeout %d secs\n", p->msg_wait_puback.msg_id,
+                (k_uptime_get_32() - p->msg_wait_puback.start_ticks)/1000);
+
+            msg_id = p->msg_wait_puback.msg_id;
+            msg_wait_puback_list_delete(msg_id);
+
+            if(config->puback_timeout_cfg.cb){
+                config->puback_timeout_cfg.cb(msg_id);
+            }
+            p = msg_wait_puback_list_head.p_next;
+
+        } else {
+            p = p->p_next;
+        }
+    }
+}
 
 static unsigned int get_next_message_id() {
+
     if (rolling_message_id + 1 >= UINT_MAX) {
```

Comment:
message ID is uint32_t and UINT_MAX is max for unsigned int. I think that the max should be checked against UINT32_MAX (if it exists), or 0xFFFFFFFF, or the message id type is changed to unsigned long. Personally, I would prefer unsigned long for compatibility, but it's up to you.

Reply:
Hi Nik,

I did not change this code. This is the original.